Web Inspector: Fix issues with go to property arrow in Computed CSS pane when details sidebar is unified

mirnovov opened the pull request:
#### b8288eb0b2ed4be5bac8f8a4765a751cf62b2e0b
<pre>
Web Inspector: Fix issues with go to property arrow in Computed CSS pane when details sidebar is unified
<a href="https://bugs.webkit.org/show_bug.cgi?id=272876">https://bugs.webkit.org/show_bug.cgi?id=272876</a>

Reviewed by Devin Rousso.

When using the Elements tab of the Inspector, the Computed pane has a button to jump to
the respective CSS property. This works well with the default two-column view, but when
&quot;show independent Styles sidebar&quot; is disabled, it can fail to scroll the respective CSS
into view or replace it with white space.

This commit fixes this problem:
- When opening details panes, do not reset the scroll position.
- Fix a bug that meant that focused/highlighted properties failed to update the layout when
  being initialised (this also can happen for other reasons,
  but is outside the scope of this commit).
- Ensure that all previous properties are always deselected, as occurs already in the
  two-column view.

Combined changes:
* Source/WebInspectorUI/UserInterface/Views/GeneralStyleDetailsSidebarPanel.js:
(WI.GeneralStyleDetailsSidebarPanel.prototype.layout):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationEditor.js:
(WI.SpreadsheetCSSStyleDeclarationEditor):
(WI.SpreadsheetCSSStyleDeclarationEditor.prototype.initialLayout):
(WI.SpreadsheetCSSStyleDeclarationEditor.prototype.selectProperties):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.deselectProperties):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.scrollToSectionAndHighlightProperty):
* Source/WebInspectorUI/UserInterface/Views/StyleDetailsPanel.js:

Canonical link: <a href="https://commits.webkit.org/278686@main">https://commits.webkit.org/278686@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4c83291a0ebac41e25a81ab060d3375efe73520

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50574 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29871 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1264 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41228 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27516 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43552 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22338 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/811 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8952 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55422 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25674 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/776 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48642 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26933 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43700 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47695 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11217 "Built successfully and passed tests") | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27798 "Hash c4c83291 for PR 27503 does not build (failure)") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26665 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsSidebarPanel.js:
```diff
@@ -40,6 +40,8 @@ WI.ComputedStyleDetailsSidebarPanel = class ComputedStyleDetailsSidebarPanel ext
-        styleRulesPanel.panel.scrollToSectionAndHighlightProperty(property);
+        setTimeout(() => {
+            styleRulesPanel.panel.scrollToSectionAndHighlightProperty(property);
+        });
```
`WI.SpreadsheetRulesStyleDetailsPanel.prototype.computedStyleDetailsPanelShowProperty` already has logic that should handle if it's not yet constructed (or in the middle of a layout).  is there some reason that that doesn't work (anymore)?

my guess is that inside `WI.SpreadsheetRulesStyleDetailsPanel.prototype.nodeStylesRefreshed` we clear `_propertyToSelectAndHighlight`, but it's possible that even then we may still `!this.isAttached || this.layoutPending`, so id probably change that to be something like this
```js
if (this._propertyToSelectAndHighlight) {
    let property = this._propertyToSelectAndHighlight;
    this._propertyToSelectAndHighlight = null;
    this.scrollToSectionAndHighlightProperty(property);
}
```

id much prefer solving this there rather than expecting every caller to use `setTimeout`/`requestAnimationFrame` :)

mirnovov replied:
Sorry, I didn't realise this scenario was already intended to be handled. I've found the root causes of this behaviour and fixed those instead.

> my guess is that inside WI.SpreadsheetRulesStyleDetailsPanel.prototype.nodeStylesRefreshed we clear _propertyToSelectAndHighlight, but it's possible that even then we may still !this.isAttached || this.layoutPending, so id probably change that to be something like this

Unless there's something I'm missing, that doesn't do anything. The inspector correctly determines the property to select (at least according to my debugger breakpoints), there's just a variety of factors that stop it actually jumping to it.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/GeneralStyleDetailsSidebarPanel.js:
```diff
@@ -69,7 +71,12 @@ WI.GeneralStyleDetailsSidebarPanel = class GeneralStyleDetailsSidebarPanel exten
     }
-
+    
+    styleDetailsPanelPreserveScroll()
```
this should go in the `// StyleDetailsPanel delegate` section just below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/GeneralStyleDetailsSidebarPanel.js:
```diff
@@ -86,7 +93,11 @@ WI.GeneralStyleDetailsSidebarPanel = class GeneralStyleDetailsSidebarPanel exten
+        if (!this._preserveScroll)
+            this.contentView.element.scrollTop = 0;
+        else
+            this._preserveScroll = false;
```
i wonder if we actually even need this anymore.  seems like a legacy behavior back from when the pseudo-class toggles were "above" the main content of the Styles sidebar panel

mirnovov replied:
If this is removed, the scroll position is saved when the user changes tabs (or switches to another sidebar pane) and then returns. Personally I would find this an improvement, so I'd be happy to make that change, but others may disagree.

dcrousso replied:
yeah im definitely in favor of that.  im pretty sure that was the behavior in the past.  not sure when that changed

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -207,11 +207,19 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
             this._element.scrollIntoView();
+            
+            if (this._delegate)
+                this._delegate.spreadsheetCSSStyleDeclarationSectionPreserveScroll();
```
you can simplify this to just be
```js
this._delegate?.spreadsheetCSSStyleDeclarationSectionPreserveScroll();
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -207,11 +207,19 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
         return false;
     }
+    
+    deselectProperties() {
```
Style: `{` should go on its own line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
```diff
@@ -90,8 +90,8 @@ WI.SpreadsheetRulesStyleDetailsPanel = class SpreadsheetRulesStyleDetailsPanel e
         for (let section of this._sections) {
-            if (section.highlightProperty(property))
-                return;
+            section.deselectProperties();
+            section.highlightProperty(property);
```
rather than call `deselectProperties` here i wonder if instead we should have `WI.SpreadsheetCSSStyleDeclarationEditor.prototype.highlightProperty` call this if nothing matches (i.e. right before the final `return false;`) as that would solve this bug and any like that also exist today or are accidentally added in the future

mirnovov replied:
This breaks when the old selection and new selection are under the same rule. I'd have to add it for both branches of `WI.SpreadsheetCSSStyleDeclarationEditor.prototype.highlightProperty` to work.

Also in the future there may be other features where the property is highlighted but the selection isn't invalidated.

dcrousso replied:
what exactly do you mean by "under the same rule"?  can you give an example?

mirnovov replied:
I.e. within the same CSS rule set or `section`

```css
.foo {
    color: blue;
    background: red;
}

.bar {
    border: 1px solid green;
}
```

If I select `color` and then go to the computed tab and click on the goto arrow for `border`, it will work, but if I click on the goto arrow next to `background` it will still break

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/StyleDetailsPanel.js:
```diff
@@ -109,6 +109,12 @@ WI.StyleDetailsPanel = class StyleDetailsPanel extends WI.View
+    spreadsheetCSSStyleDeclarationSectionPreserveScroll()
+    {
+        if (this._delegate)
+            this._delegate.styleDetailsPanelPreserveScroll();
```
you can simplify this to just be
```js
this._delegate?.styleDetailsPanelPreserveScroll();
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationEditor.js:
```diff
@@ -60,7 +61,13 @@ WI.SpreadsheetCSSStyleDeclarationEditor = class SpreadsheetCSSStyleDeclarationEd
-        this.element.addEventListener("focus", () => { this.focused = true; }, true);
+        this.element.addEventListener("focus", () => {
+            if (this._selectionFocusOverride)
+                this._selectionFocusOverride = false;
```
can you explain more why this is needed?  is it that `this.editing` is now `true`?  perhaps the issue is with `this._style.locked`?

mirnovov replied:
The `layout()` function of this class terminates when `editing` is true. Until this commit, `_focused` was always set to true, as usually focusing meanings that we are editing the CSS and we shouldn't layout. However focusing is _also_ done when an item is highlighted (which AFAIK is either when it is arrived at from a go to arrow or other UI element, or when it is selected manually by the user by clicking in the row but outside the text field). 

So when the go to arrow was clicked, it would prevent the initial layout and therefore all of the properties could appear blank. This didn't occur with two-column view since the styles pane was already layouted.

`this._style.locked` doesn't seem to be checked as part of this initial condition in the `layout()` function, so I don't think that's the cause.

dcrousso replied:
hmm that makes me think the real issue is that we consider highlighting as a form of focus, but i can see how it might be difficult/annoying to untangle those two concepts 🤔

maybe we could have the `"focus"` event handler only do something if it's user initiated (i.e. `isTrusted`)?

is the issue only on the first `layout` or in any `layout`?  if the former, perhaps we instead just adjust the `this.focus` check inside `layout` to also check for `this.didInitialLayout`?

mirnovov replied:
Yeah, I did think the same thing - that ideally focus wouldn't do double-duty for two disparate things, but it's probably not something to do rn. I also think there may be AX concerns with unbundling it, given that both of them are "focusing" in an accessibility sense.

I have already tried `this.didInitialLayout` but it's not false when necessary (since it's changed after `initialLayout` but before the first `layout` in `View`). Similarly `isTrusted` is always true.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/StyleDetailsPanel.js:
```diff
@@ -109,7 +109,7 @@ WI.StyleDetailsPanel = class StyleDetailsPanel extends WI.View
         // Implemented by subclasses.
     }
-
+    
```
oops, please undo :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationEditor.js:
```diff
@@ -391,6 +398,7 @@ WI.SpreadsheetCSSStyleDeclarationEditor = class SpreadsheetCSSStyleDeclarationEd
         }
 
         this._suppressBlur = true;
+        this._selectionFocusOverride = true;
```
if we truly do need this maybe we could leverage `this._suppressBlur` instead?
```js
this.element.addEventListener("focus", () => {
    if (!this._suppressBlur)
        this.focused = true;
}, true);
```

mirnovov committed:
Web Inspector: Fix issues with go to property arrow in Computed CSS pane when details sidebar is unified
https://bugs.webkit.org/show_bug.cgi?id=272876

Reviewed by Devin Rousso.

When using the Elements tab of the Inspector, the Computed pane has a button to jump to
the respective CSS property. This works well with the default two-column view, but when
"show independent Styles sidebar" is disabled, it can fail to scroll the respective CSS
into view or replace it with white space.

This commit fixes this problem:
- When opening details panes, do not reset the scroll position.
- Fix a bug that meant that focused/highlighted properties failed to update the layout when
  being initialised (this also can happen for other reasons,
  but is outside the scope of this commit).
- Ensure that all previous properties are always deselected, as occurs already in the
  two-column view.

Combined changes:
* Source/WebInspectorUI/UserInterface/Views/GeneralStyleDetailsSidebarPanel.js:
(WI.GeneralStyleDetailsSidebarPanel.prototype.layout):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationEditor.js:
(WI.SpreadsheetCSSStyleDeclarationEditor):
(WI.SpreadsheetCSSStyleDeclarationEditor.prototype.initialLayout):
(WI.SpreadsheetCSSStyleDeclarationEditor.prototype.selectProperties):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.deselectProperties):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.scrollToSectionAndHighlightProperty):
* Source/WebInspectorUI/UserInterface/Views/StyleDetailsPanel.js:

Canonical link: https://commits.webkit.org/278686@main
